packages/vm/src/imports.rs

```diff
@@ -150,19 +147,29 @@ macro_rules! write_region {
 }
 
 /// Reads a storage entry from the VM's storage into Wasm memory
-pub fn do_read<S: Storage, Q: Querier>(
-    ctx: &mut Ctx,
-    key_ptr: u32,
-    value_ptr: u32,
-) -> VmResult<i32> {
-    let key = read_region!(ctx, key_ptr, MAX_LENGTH_DB_KEY);
+pub fn do_read<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResult<u32> {
+    let key = read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY)?;
     // `Ok(expr?)` used to convert the error variant.
     let value: Option<Vec<u8>> =
         with_storage_from_context::<S, Q, _, _>(ctx, |store| Ok(store.get(&key)?))?;
-    Ok(match value {
-        Some(buf) => write_region!(ctx, value_ptr, &buf),
-        None => errors::read::KEY_DOES_NOT_EXIST,
-    })
+
+    let out_data = match value {
+        Some(data) => data,
+        None => return Ok(0),
+    };
+
+    let out_ptr = with_func_from_context::<S, Q, u32, u32, _, _>(ctx, "allocate", |allocate| {
+        let out_size = to_u32(out_data.len())?;
+        let ptr = allocate.call(out_size)?;
+        if ptr == 0 {
+            return Err(make_generic_err(
+                "Got a null Wasm pointer from allocate(). This is not allowed by CosmWasm even if it is valid Wasm. The contract needs to ensure to allocate at a non-null address. See https://stackoverflow.com/q/62042733/2013738 for more information.",
```

Comment:
You can split this string over multiple lines without changing its contents:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ba2a21df978d833fb4f41c84faec5db3

Reply:
This is outdated after the creation of a structured error for this case in 86202113b6d420740382821ba586a74b58db59da.

I want to keep the check as close as possible to the `allocate` call, since this is now disallowed to return 0 by convention.